eos-split-flatpak-repo

```diff
@@ -0,0 +1,307 @@
+    _, sysroot_locked = sysroot.try_lock()
+    if not sysroot_locked:
+        raise SplitError(f'Could not lock sysroot {root}')
+
+    ostree_repo = sysroot.repo()
+    set_repo_path_attr(ostree_repo)
+    logger.debug('OSTree repo path: %s', ostree_repo.path)
+
+    ostree_dir_path = os.path.dirname(ostree_repo.path)
+    var_path = os.path.join(ostree_dir_path, 'deploy/eos/var')
+    logger.debug('EOS /var path: %s', var_path)
+    flatpak_dir_path = os.path.join(var_path, 'lib/flatpak')
+    logger.debug('Flatpak dir path: %s', flatpak_dir_path)
+    flatpak_repo_path = os.path.join(flatpak_dir_path, 'repo')
+    logger.debug('Flatpak repo path: %s', flatpak_repo_path)
+    if not os.path.samefile(ostree_repo.path, flatpak_repo_path):
+        logger.info('OSTree repo and Flatpak repo have already been split')
+        return False
+
+    # Lock the current repo
+    with repo_lock(ostree_repo, 'exclusive'):
+        # Create the new flatpak repo
+        new_flatpak_repo_path = os.path.join(
+            os.path.dirname(ostree_repo.path), 'new-flatpak-repo')
+        logger.info('Creating new Flatpak repo at %s', new_flatpak_repo_path)
+        duplicate_dir(ostree_repo.path, new_flatpak_repo_path)
+        new_flatpak_repo = OSTree.Repo.new(
+            Gio.File.new_for_path(new_flatpak_repo_path))
+        new_flatpak_repo.open()
+        set_repo_path_attr(new_flatpak_repo)
+
+        # Lock the new repo
+        with repo_lock(new_flatpak_repo, 'exclusive'):
+            # Gather the remotes
+            logger.info('Gathering OS and Flatpak remotes')
+            os_remotes, flatpak_remotes = gather_remotes(ostree_repo)
+
+            # Gather the refs
+            logger.info('Gathering OS and Flatpak refs')
+            os_refs, flatpak_refs, _ = gather_refs(ostree_repo, os_remotes,
+                                                   flatpak_remotes)
+
+            # Remove OS remotes and refs from the flatpak repo
+            remove_remotes(new_flatpak_repo, RepoType.OS, os_remotes)
+            remove_refs(new_flatpak_repo, RepoType.OS, os_refs)
+            if prune:
+                logger.info('Pruning Flatpak repo %s', new_flatpak_repo_path)
+                new_flatpak_repo.prune(OSTree.RepoPruneFlags.REFS_ONLY, -1)
+
+            # Convert the new flatpak repo to a bare-user-only repo.
+            logger.info('Changing Flatpak repo %s mode to bare-user-only',
+                        new_flatpak_repo_path)
+            flatpak_repo_config = new_flatpak_repo.copy_config()
+            flatpak_repo_config.set_value('core', 'mode', 'bare-user-only')
+            new_flatpak_repo.write_config(flatpak_repo_config)
+
+            # Swap the new flatpak repo into place.
+            logger.info('Replacing Flatpak repo %s with %s',
+                        flatpak_repo_path, new_flatpak_repo_path)
+            replace_symlink(flatpak_repo_path, new_flatpak_repo_path)
+
+            # Most likely var/lib/flatpak is a symlink to
```

Comment:
A failure after this point will not be solvable by re-running the script (since the flatpak repo will have been separated by this point). Should the code below try a bit harder to carry on in the face of error (for example, pruning the OSTree repo even if the `flatpak_dir_path` symlink can’t be fixed)?

Overall, what’s the error handling strategy for the script? Mostly it seems to be relying on everything being atomic on the `replace_symlink()` call above, which seems reasonable. Might be good to have that documented.

I can’t see a way to make the code below fit in with the above atomic switch to make it inherently safe.

Reply:
Yeah, I really couldn't figure out what to do here since it's orthogonal but related to the actual repo split. Everything will continue to work fine if `/var/lib/flatpak` is a symlink to `/sysroot/flatpak`, but it's not clean. And currently if there's an error at this point rerunning the script won't address it since it only checks if `/var/lib/flatpak/repo` is a symlink.

I guess there should really be 2 distinct checks and actions:

* Check if `/var/lib/flatpak/repo` is a symlink and split the repos if it is
* Check if `/var/lib/flatpak` is a symlink and replace the symlink with the real flatpak dir

This also affects the systemd unit, which I haven't tested and currently only has `ConditionPathIsSymbolicLink=/var/lib/flatpak/repo`. I believe that would have to be updated to:
```
ConditionPathIsSymbolicLink=|/var/lib/flatpak
ConditionPathIsSymbolicLink=|/var/lib/flatpak/repo
```

Reply:
> I guess there should really be 2 distinct checks and actions:

Sounds reasonable to me.